src/storage/db_interface_common.py

```diff
@@ -310,3 +311,50 @@ def release_unpacking_lock(self, uid):
 
     def drop_unpacking_locks(self):
         self.main.drop_collection('locks')
+
+    def _collect_analysis_tags_from_children(self, uid):
+        children = self._fetch_children_with_tags(uid)
+        unique_tags = {}
+        for child in children:
+            self._collect_analysis_tags(child, unique_tags)
+        return unique_tags
+
+    def _collect_analysis_tags(self, file_object, analysis_tags):
+        for name, analysis in [(n, a) for n, a in file_object.processed_analysis.items() if 'tags' in a]:
+            if 'file_system_flag' in analysis and analysis['file_system_flag']:
+                analysis = self.retrieve_analysis(file_object.processed_analysis, analysis_filter=[name, ])[name]
+
+            for tag_type, tag in analysis['tags'].items():
+                if tag_type != 'root_uid' and tag['propagate']:
+                    append_unique_tag(analysis_tags, tag, name, tag_type)
+
+    def _fetch_children_with_tags(self, uid):
+        plugins_with_tag_propagation = [  # FIXME This should be inferred in a sensible way. This is not possible yet.
+            'crypto_material', 'cve_lookup', 'known_vulnerabilities', 'qemu_exec', 'software_components',
+            'users_and_passwords'
+        ]
+        uids = set()
+        for plugin in plugins_with_tag_propagation:
+            uids.update(
+                set(
+                    get_list_of_all_values(
+                        self.file_objects,
+                        '$_id',
+                        match={
+                            'virtual_file_path.{}'.format(uid): {'$exists': 'true'},
+                            f'processed_analysis.{plugin}.tags': {'$exists': 'true'}
+                        }
+                    )
+                )
+            )
+        return self.get_objects_by_uid_list(uids, analysis_filter=plugins_with_tag_propagation)
+
+
+def append_unique_tag(unique_tags: Dict[str, dict], tag: dict, plugin_name: str, tag_type: str) -> None:
+    if plugin_name in unique_tags:
+        if tag_type in unique_tags[plugin_name] and tag not in unique_tags[plugin_name].values():
+            unique_tags[plugin_name][f'{tag_type}-alt-{random.randint(0, 1000)}'] = tag
```

Comment:
Is this necessary for multiple results of the same kind?
I don't like the random(9999) solutions too much we could also use the length:
```suggestion
            unique_tags[plugin_name][f'{tag_type}-alt-{len(unique_tags[plugin_name])}'] = tag
```

Reply:
This does not solve anything. Two _OS version_ tags would still generate the same key when they e.g. have values _3.10.20_ and _3.10.35_. In case you only have two it works due to the `-alt-` but for three linux versions you overwrite a value. Dicts are not hashable, aye? The _safe_ but nasty solution is an iterating while / break to identify the next unused key starting at _-alt-0_.